### PR TITLE
[build] Bump minimum compiler versions and enable C++14

### PR DESCRIPTION
**Pull request by kkm000**
```
===================================
                Min. req. version:
Compiler       Current       New
===========    =======     ========
GCC             4.9.1   >    5.0
LLVM Clang      3.3     >    3.5
XCode Clang     5.0     >    6.0
===================================
```
At the same time, upgrade OpenFST to 1.7.2 (which I've been using for quite a while without a glitch).

/cc @hugovbraun

Close: #4297
Close: #4245

### CODE REVIEW

File: tools/Makefile [hunk 1]
```diff
@@ -19,12 +17,13 @@ SCTK_GITHASH = 20159b5
 
 # Default features configured for OpenFST; can be overridden in the make command line.
 OPENFST_CONFIGURE ?= --enable-static --enable-shared --enable-far \
-                     --enable-ngram-fsts --enable-lookahead-fsts --with-pic
+                     --enable-ngram-fsts --enable-lookahead-fsts \
+                     --disable-dependency-tracking --with-pic
```
Comment (galv) [inline tools/Makefile:21]: Is `--disable-dependency-tracking` really a good idea? You are basically assuming that users won't modify and develop openfst source code, which is something I've done in the past.

File: tools/Makefile [hunk 2]
```diff
@@ -78,13 +77,13 @@ else ifeq ($(OS),Windows_NT)
...
 else
-  openfst_add_CXXFLAGS = -g -O2
+  openfst_add_CXXFLAGS = -g -O3
 endif
 
 openfst-$(OPENFST_VERSION)/Makefile: openfst-$(OPENFST_VERSION)
 	cd openfst-$(OPENFST_VERSION)/ && \
 	./configure --prefix=`pwd` $(OPENFST_CONFIGURE) CXX="$(CXX)" \
-		CXXFLAGS="$(CXXFLAGS) $(openfst_add_CXXFLAGS)" \
+		CXXFLAGS="$(openfst_add_CXXFLAGS) $(CXXFLAGS)" \
```
Comment (galv) [inline tools/Makefile:86]: Given that you've made this change to swap the order of `openfst_add_CXXFLAGS` and `CXXFLAGS`, can you change -O3 back to -O2 in `openfst_add_CXXFLAGS`? Right now, a semi-large change is adding different, unrelated changes. If you really want -O3 for your use case, you can do `CXXFLAGS="-O3" make`

**Review by galv — COMMENTED** [review]
excellent forward progress

File: tools/Makefile (hunk 1 above)
Comment (kkm000) [reply tools/Makefile:21]: When automatically compiling it as part of Kaldi, I think it's a good idea. If you really hack OpenFST in context of Kaldi to the point dependencies change, you'll probably rerun configure anyway many times.

File: tools/Makefile (hunk 2 above)
Comment (kkm000) [reply tools/Makefile:86]: Good point! Done.

File: tools/Makefile (hunk 1 above)
Comment (jtrmal) [reply tools/Makefile:21]: I think the --disable-dependency-tracking  flag should be removed. I don't want to optimize for the time of machine. I want to optimize for time of people who are trying to track down an issue.

**Comment by jtrmal** [comment]
LGTM, with the exception of the tracking dependecies flag

File: tools/Makefile (hunk 1 above)
Comment (kkm000) [reply tools/Makefile]: Done.

**Comment by kkm000** [comment]
@jtrmal, ok, I probably underestimate the importance of dependencies. Anyway, the build with libtool is so slow that the speedup from skipping 'make depend' is negligible anyway. FWIW, later versions of OpenFST come with Bazel build files, but no CMake :( The library (w/o bins) builds is about 6-7 seconds with CMake/ninja if `-j` arg (and the # of hyperthreads) equals or exceeds the number of `.cc` files in `src/lib`. And, alas, there is no way to build it with `-std=c++14`. When we upgrade the whole rig to CMake, providing our own CMake build framework for OpenFST seems to be the best approach (given that you wrote it already for the Windows port, it's a zero effort task, with a sizable build time benefit).

I'll merge when Travis completes, thanks for looking!
